### PR TITLE
Expose objective functions to the Python interface.

### PR DESCRIPTION
**Pull request by trivialfis**
Ref https://github.com/dmlc/xgboost/issues/7693 https://github.com/dmlc/xgboost/issues/9043

- Expose all objective functions to the Python interface.
- No change in the `Learner` yet. My plan is to make an independent C API to create objectives with the `split_grad` customization. This will have to wait until the plan is clear.
- Metrics are not exposed. I'm focusing on getting vector-leaf to work with built-in objectives.
- Remove the `TreeObjective`, since the linear model is now deprecated.
- `cv` function is not supported yet.
- Naming of the parameters are described in the referenced issue 

I named the classes `RegSquaredError` instead of `SquaredError` to maintain consistency with the C++ naming. But I'm open to suggestions.

### CODE REVIEW

**Review by RAMitchell — COMMENTED** [review]
Exposing objectives is great - but the split gradient method doesn't make sense to me. It implies that you would want a different method of approximating gradients for each objective type, but it seems the same techniques apply (SVD, random projection).

**Comment by trivialfis** [comment]
The alternative is an independent reducer as sketch boost, which I did took during initial implementation.

My final choice of using the objective as the vehicle to deliver the feature as I saw the technique as using different loss functions for split and leaf values, instead of using the same loss with dimension reduction applied on the gradient. This view was briefly mentioned in the document. The split loss may or may not be derived from dimension reduction techniques.

I'm open to change if the view is considered not productive. The interface is marked as wip, we can change.

Then, it's not difficult to compose either, just define a reducer with the `split_grad` as a mixin for your objective class.

**Comment by RAMitchell** [comment]
> My final choice of using the objective as the vehicle to deliver the feature as I saw the technique as using different loss functions for split and leaf values, instead of using the same loss with dimension reduction applied on the gradient. This view was briefly mentioned in the document. The split loss may or may not be derived from dimension reduction techniques.

Do you have an example or reference for this?

If you want I can help implement a simple random projection then we have just one parameter.

**Comment by trivialfis** [comment]
> Do you have an example or reference for this?

It's a perspective. You can **see** the gradient derived from dimension reduction as a different loss gradient: $\partial L_{split}$, $\partial L_{leaf}$, there's no need to put a strict link between them that says dimension reduction must be employed to obtain one from the other. To come up with a simple example for this [demo](https://github.com/dmlc/xgboost/blob/a1e4e74ba8d0d6c18dad58e7745177777a1538ce/demo/guide-python/multioutput_reduced_gradient.py#L28):
``` python
class LsObjSum(LsObjMean):
    def __init__(self, device: str) -> None:
        super().__init__(device=device)

    def __call__(
        self, iteration: int, y_pred: np.ndarray, dtrain: xgb.DMatrix
    ) -> Tuple[np.ndarray, np.ndarray]:
        self.dtrain = dtrain
        self.y_pred = y_pred
        return super().__call__(iteration, y_pred, dtrain)
    def split_grad(
        self, iteration: int, grad: np.ndarray, hess: np.ndarray
    ) -> Tuple[np.ndarray, np.ndarray]:
       # Derive the gradient from the mean of the output values instead of using the mean of the gradient
        y = self.dtrain.get_label()
        y_pred = self.y_pred
        y_mean = np.mean(y, axis=1)
        p_mean = np.mean(y_pred, axis=1)
        grad = p_mean - y_mean
        hess = np.ones(grad.shape)
        return grad, hess
```

> If you want I can help implement a simple random projection then we have just one parameter.

I'm not sure I want to make random projection the only choice, even if I drop the `split_grad` view. That seems like an unnecessary restriction, out of so many reduction techniques out there. I don't think this simplifies or enhances XGBoost.

**Comment by trivialfis** [comment]
I reverted the changes for the vector-leaf. This PR now only handles exposing the objective functions.

**Comment by RAMitchell** [comment]
This split gradient above is always an underestimate of split gain due to this inequality
image

Random projection is unbiased and will converge in expection, always, for every loss. Doesn't it just solve the problem?

**Comment by trivialfis** [comment]
> This split gradient above is always an underestimate of split gain due to this inequality

It's just an example that one can see the gradient this way.

> Random projection is unbiased and will converge in expection, always, for every loss. Doesn't it just solve the problem?

It's certainly one of the candidates. But my point is, why put in this restriction?

**Comment by trivialfis** [comment]
@RAMitchell I have reverted all changes related to the `split_grad`, this PR now only handles exposing the objective. Please help take another look when you are available. Some notes have been added to the PR description.

File: python-package/xgboost/testing/objective.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+from .data import get_cancer
+from .utils import Device
+
+
+def check_train_regression_objectives(device: Device) -> None:
+    """Test training with regression objective classes."""
+    X, y, _ = make_regression(100, 5, use_cupy=device == "cuda")
+    dm = DMatrix(X, label=y)
+
+    for obj_inst, obj_name in [
```
Comment (RAMitchell) [inline python-package/xgboost/testing/objective.py:48]: I guess we can't use pytest parameterize with this organisation of tests?

In legate-boost I had parameterisation over device and objectives.

File: python-package/xgboost/testing/objective.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        (RankPairwise(), "rank:pairwise"),
+        (RankMAP(), "rank:map"),
+    ]:
+        bst = train({"device": device}, dm, num_boost_round=5, obj=obj_inst)
+        cfg = json.loads(bst.save_config())
+        assert cfg["learner"]["objective"]["name"] == obj_name
+        assert obj_inst.name == obj_name
+
+
+def check_equivalence(device: Device) -> None:
```
Comment (RAMitchell) [inline python-package/xgboost/testing/objective.py]: Would be nice to see this parameterised over objectives.

File: python-package/xgboost/core.py [hunk 3]
```diff
@@ -2280,21 +2278,14 @@ def train_one_iter(grad: NumpyOrCupy, hess: NumpyOrCupy) -> None:
         vgrad: Optional[ArrayLike]
```
Comment (RAMitchell) [inline python-package/xgboost/core.py:2278]: Just curious what the latency differences between using the python interface or internal. One possible simplification could be to always use the python, there is just one code path.

**Review by RAMitchell — APPROVED** [review]
LGTM. I assume documentation comes at a later stage? We will probably have to render all the math formulas in the docs.

File: python-package/xgboost/testing/objective.py (hunk 1 above)
Comment (trivialfis) [reply python-package/xgboost/testing/objective.py:48]: There are a couple of options, like parametrization, pytest subtests, or a simple loop. I don't have a strong preference for this specific test, as it's unlikely we need to isolate the tests to reproduce the issues. If one of them fails, all of the objectives should fail.

File: python-package/xgboost/core.py (hunk 3 above)
Comment (trivialfis) [reply python-package/xgboost/core.py:2278]: The internal assumes split gradient must be available, and only does an extra leaf value computation at the end of iteration if there's an extra leaf value gradient. This is easy to implement as we only need one extra step.

But that's not intuitive to users since the algorithm creates an extra "split gradient", at least that's the mental model. So, the interface assumes a value gradient is available, as in normal gradient boosting. The assumption is switched here.

**Comment by trivialfis** [comment]
> We will probably have to render all the math formulas in the docs.

I'm thinking if it's possible to reuse the document instead of copying and pasting it into Python files.

File: python-package/xgboost/testing/objective.py (hunk 2 above)
Comment (trivialfis) [reply python-package/xgboost/testing/objective.py]: Done.

File: python-package/xgboost/core.py (hunk 3 above)
Comment (trivialfis) [reply python-package/xgboost/core.py:2278]: I misread the original question. There's no difference in latency, it's just setting parameter. Yes, I think a single code path would be super nice.
